File: python/rabbit_listener.py
#!/usr/bin/env python2

import pika
import json

# Setup the rabbit connection and queue
rabbit_conn = pika.BlockingConnection(
    pika.ConnectionParameters(host='localhost', port=5672),
)
exchange = 'rethinkdb'
channel = rabbit_conn.channel()
channel.exchange_declare(exchange=exchange, type='topic', durable=False)
queue = channel.queue_declare(exclusive=True).method.queue

# Bind to all changes on the 'mytable' topic
channel.queue_bind(queue, exchange, routing_key='mytable.*')

# Listen for changes and print them out
print 'Started listening...'

for method, properties, payload in channel.consume(queue):
    change = json.loads(payload)
    change_typ = method.routing_key.split('.')[1]
    print 'RabbitMQ -(', method.routing_key, ')-> Listener'
    if change_typ == 'create':
        print json.dumps(change['new_val'], indent=True, sort_keys=True)
    elif change_typ == 'delete':
        print json.dumps(change['old_val'], indent=True, sort_keys=True)
    else:
        print 'Old value:'
        print json.dumps(change['old_val'], indent=True, sort_keys=True)
        print 'New value:'
        print json.dumps(change['new_val'], indent=True, sort_keys=True)
    print '='*80, '\n'
    

File: javascript/rabbit_listener.js
#!/usr/bin/env node

var amqp = require('amqplib');

// Setup the rabbit connection

var rabbitConn;
var channel;
var exchange = 'rethinkdb';
var queue;

// Create the rabbit connection
amqp.connect('amqp://localhost:5672').then(function(conn){
    rabbitConn = conn;
    // Create the rabbit channel
    return rabbitConn.createChannel();
}).then(function(ch){
    channel = ch;
    // Create the exchange (or do nothing if it exists)
    return channel.assertExchange(exchange, 'topic', {durable: false});
}).then(function(){
    // Create the queue
    return channel.assertQueue('', {exclusive: true});
}).then(function(q){
    queue = q.queue;
    // Bind the queue to all topics about 'mytable'
    return channel.bindQueue(queue, exchange, 'mytable.*');
}).then(function(){
    console.log('Started listening...');
    channel.consume(queue, function(msg){
        // Handle each message as it comes in from RabbitMQ
        var change = JSON.parse(msg.content);
        var tablename = msg.fields.routingKey.split('.');
        console.log(tablename, '-> RabbitMQ -(',
                    msg.fields.routingKey, ')-> Listener');
        console.log(JSON.stringify(change, undefined, 2));
        console.log(new Array(81).join('=') + '\n')
    })
})
